### ISSUE TITLE
Error: [401] - "Unauthorized" - JQL: project='AI-Agemni' (line 74) - shared sheet with many users.

### ISSUE THREAD

**Issue by jeffrmanley** [ljay79/jira-tools#188]
Several Users share a gSheet, most users have the Project Aid for Jira setup under their user. I'm attempting to use this:

=JST_getTotalForSearchResult("project='AI-Agemni'") 

which produces the error Error: [401] - "Unauthorized" - JQL: project='AI-Agemni' (line 74). 

It was working when I first created the column but it is no longer working. From another tab I'm still able to run Add-ons --> Project Aid for Jira --> List issues from filter 

The Temp User = AANBUuVCaxK4Nz55qGPbx10PJomi0hhvlxx8pq7Zx0dI4p0qdj2lcEn3PMVGGqSvy99lg18h5Bjf

You should be able to see a few attempts where the error was produced in addition to the successful run from the List Jira issues from filter (using the IVP-CurrentProjects filter)

**Comment by ljay79** [comment]
Thx @jeffrmanley for reporting this.
I will certainly have a look into it asap.

**Comment by ljay79** [comment]
Are you able to forcefully reproduce the issue in a completly new Google Spreadsheet and can then provide the steps for me to try myself reproducing the issue.
Without having me able to reproduce it, this is an hard issue to research and/or fix.

**Comment by ljay79** [comment]
What i notice already so far, the custom function _JST_getTotalForSearchResult_ appears to being executed in a different user scope.
Your successfuly calls with "List issues from filter" appear under UserKey: _"AANBU*****GqSvy99lg18h5Bjf"_
While the failed calls to the custom functions are logged under UserKey: _"AANBU*****FcRqD5w39HZ5Q"_

Im trying to analyize this in my environment as best as i can. Sadly im a little limited as i run a single user Jira instance atm :)   Im not a pro user...

**Comment by jeffrmanley** [comment]
I've uninstall and reinstalled the Add-On. The new Temp User is AANBUuXNismlhcD+t9PM+7mpDK0BGVZPDJjwIlJzWkmkwOLjMNu5D6JQAiQ7/pAOe8gfrgSkc8YU

and I'm still experiencing the same [401] - "Unauthorized" error

Thanks for helping to get this solved!!

**Comment by jeffrmanley** [comment]
Another possible important note. We have some users in this gSheet setup in the Add-On to pull data from a different Jira Server.

**Comment by ljay79** [comment]
Hmmm...   the custom function you are getting an error is added into the cell by yourself or by a different user?

**Comment by ljay79** [comment]
@paul-lemon maybe you can replicate this issue with your multi-user setup?

**Comment by ljay79** [comment]
Damn, not sure how i could forget about this :/

Im so sorry.
**So in fact, google's custom functions do use the document owner's configured site and Jira account to query Jira, when sharing a spreadsheet anyone who can edit it will be able to use this function and import Jira data using your account.**

I must find a way, of highlighting / warning users upfront about this.
After long research and reviewing my notes (and comparing to other add-ons), there is actually no work arround this limitation.

**Comment by ljay79** [comment]
I think a good way of makeing sure, users are aware of this and are warned clearly, i think about having all custom functions disabled by default and let a user enable it only via the settings dialog with a good warning of this behavior.

What to you guys think about this?

**Comment by paul-lemon** [comment]
I can test this sometime soon.

> @paul-lemon maybe you can replicate this issue with your multi-user setup?

i'll have a go.

**Comment by jeffrmanley** [comment]
> Hmmm... the custom function you are getting an error is added into the cell by yourself or by a different user?

I'm adding the formula myself.  As a matter of fact I have a couple Google Apps accounts that have access to this sheet and I have one setup to point to one Jira server and the other setup to the 2nd Jira server

> I think a good way of making sure, users are aware of this and are warned clearly, i think about having all custom functions disabled by default and let a user enable it only via the settings dialog with a good warning of this behavior.
> 
> What to you guys think about this?

Yes, I believe that disabling with clear warning and let the user agree to enable would be a good idea. 

Since these are read-only functions and we're using in reporting that everyone with access to the share sheet are allowed to view it works fine.

**Comment by jeffrmanley** [comment]
Some additional testing. I created a new gSheet with both my Google Apps users. 

User A has setup page pointing to Jira Server A
User B has setup page pointing to Jira Server B

Created this formula with User A and it worked fine: =JST_getTotalForSearchResult("project='AI-" & upper(A2) & "' and status='100%' and type in (subqb, Task) AND qb_ID !~ 0 AND qb_ID is not EMPTY")/JST_getTotalForSearchResult("project='AI-" & upper(A2) & "' and type in (subqb, Task) AND qb_ID !~ 0 AND qb_ID is not EMPTY")

Created a simple formula with User B and it gave an error 400 indicating the project name and type are invalid (however those were confirmed to be valid by doing a Jira query on Jira Server B). If I change the JST_getTotalForSearchResult formula to a valid project name on Jira Server A it works...conclusion is that this custom formula is using the settings of User A instead of User B

**Comment by ljay79** [comment]
Thanks for this extensive evaluation @jeffrmanley .
Indeed i can only confirm and assure with confidence: All custom functions do use the document owner's configured site and Jira account to query Jira.

**Comment by paul-lemon** [comment]
I made some tests and can confirm this also.
I set up sheet with some custom functions in the cells.
These displayed correct results.
I shared it with another user who then in the settings entered an invalid URL and password which did not connect to a JIRA server
They were able to still use custom functions in the sheet - both refreshing them and adding new ones.
They were not able to use other dialogues as there was an error with the setting.
